src/System.Text.Json/src/System/Text/Json/JsonReader.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Text.Parsing;
+using System.Text.Utf8;
+
+namespace System.Text.Json
+{
+    public struct JsonReader : IDisposable
+    {
+        private readonly Utf8String _str;
+        private int _index;
+
+        private static readonly List<Utf8CodeUnit> EmptyString = new List<Utf8CodeUnit>
+        {
+            new Utf8CodeUnit((byte) ' '),
+            new Utf8CodeUnit((byte) '\n'),
+            new Utf8CodeUnit((byte) '\r'),
+            new Utf8CodeUnit((byte) '\t')
+        };
+
+        private static readonly Utf8CodeUnit QuoteString = new Utf8CodeUnit((byte) '"');
```

Comment:
@krwq, it might be worth to have better APIs for this. It seems like casting chars to bytes might lead to bugs.